### DISCUSSION TITLE
Getting out of while(true) by using break; in Java

### DISCUSSION THREAD

**Discussion by gyomin0315** [community/community#54125] [Programming Help] [answered]
### Body

Hi, I was trying a exercise about getting out of while(true). I definitely used break; but my program doesn't stop and it is keep running the program and i'm not sure what is the reason. Here is the code. And also, I used Eclipse.

package myPackage;

import java.util.*;

public class Switch_and_For {

	public static void main(String[] args)
	{
		//String addr = "서울";

		System.out.println("Write the city where you live");
		
		Scanner sc = new Scanner(System.in);
		

		while(true)
		{
			
			String addr = sc.next();
			
			//int addr = sc.nextInt();
		
			
			switch(addr)
			{
			case "Seoul":
				System.out.println("I live in Seoul.");
				break;
			case "Busan":
				System.out.println("I live in Busan.");
				break;
			case "Daejeon":
				System.out.println("I live in Daejeon.");
				break;
			case "Sejong":
				System.out.println("I live in Sejong.");
				break;
			default :
				System.out.println("Write again.");
	

			}
			
		
			if(addr=="Seoul"||addr=="Busan"||addr=="Daejeon"||addr=="Sejong")
			{
				break;
				
			}

		}

		System.out.println("The end");

	}

}

// Result
/*
Write the city where you live
Seoul
I live in Seoul.
Daejeon
I live in Daejeon.
*/

The program doesn't print "The end" and it proceeding while part over and over again. 
I switched String addr into integer and the program worked well.
But I have no idea why this program doesn't come out of while part when I use addr as a String.
Your answer would be a lot help to me.

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by jge162** [comment, marked as answer]
Good day to you @gyomin0315 The reason your program doesn't come out of the while loop is that you are using the equality operator (==) to compare strings in the if statement where you check if the user input is one of the valid options ("Seoul", "Busan", "Daejeon", "Sejong"). Therefore, the string comparison using == always returns false, and your program never breaks out of the while loop.

in java you should use .equal, try this and please report back. 

```java
if(addr.equals("Seoul")||addr.equals("Busan")||addr.equals("Daejeon")||addr.equals("Sejong"))
{
    break;    
}
```

**Reply by gyomin0315** [reply]
Thank you for your reply. Your answer works perfectly. 
But i didn't fully understand why I can't use equality operator(==).
To be specific, I'm not sure why the string comparison using == always returns false.
What is the difference between addr.equals("Seoul") and addr == "Seoul" ?
Your information would be a lot help to me.
And thank you for your answer.

**Reply by jge162** [reply]
Java restriction with == you can’t use it to compare strings.

**Reply by gyomin0315** [reply]
Thank you for your answer. It helped me a lot.

**Reply by jge162** [reply]
Anytime. Have a great rest of your day @gyomin0315

**Comment by sabrysm** [comment]
The reason your program is not terminating when you use a String variable is because of how Java handles String comparison. In your code, you are using the == operator to compare the addr variable with the city names. However, this operator checks if the two variables refer to the same object in memory, rather than if their contents are the same.

In Java, the equals() method should be used to compare the contents of two String objects. So instead of using ==, you should use the equals() method to compare the contents of addr with the city names. Here's the modified code:

```
if(addr.equals("Seoul") || addr.equals("Busan") || addr.equals("Daejeon") || addr.equals("Sejong"))
{
    break;
}
```
Alternatively, you can use the equalsIgnoreCase() method to compare the contents of addr with the city names, ignoring the case of the letters:

```
if(addr.equalsIgnoreCase("Seoul") || addr.equalsIgnoreCase("Busan") || addr.equalsIgnoreCase("Daejeon") || addr.equalsIgnoreCase("Sejong"))
{
    break;
}
```
With either of these modifications, your program should terminate the while loop and print "The end" when a valid city name is entered.

**Reply by gyomin0315** [reply]
Thank you for your answer! It helped me a lot.

**Reply by sabrysm** [reply]
Anytime 😄

**Comment by RedSky-ll** [comment]
Instead of using the equality operator (==), to check if two strings are the same, use the .equals() method.

**Comment by hridesh-net** [comment]
The reason your program doesn't stop and continues running even after using break; is because of the way you are comparing strings in your code.
To fix the issue, you should modify your if statement where you are checking the value of addr to use the equals() method as follows:
```
if (addr.equals("Seoul") || addr.equals("Busan") || addr.equals("Daejeon") || addr.equals("Sejong")) {
    break;
}
```
